### ISSUE TITLE
TFLite float16 weight quantization causes >0.01 output divergence for attention with tanh softcapping

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#117541]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.21

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.11

### Current behavior?

TensorFlow Lite produces different results from Keras fp32 for an attention-style computation after float16 weight quantization.

Specifically, the model computes Q/K/V projections, scaled dot-product attention, tanh-based logit softcapping, softmax, and the final attention output. Keras fp32 execution produces finite results, while the converted TFLite model also produces finite results but differs beyond the tolerance.

In the reproducer below, the maximum absolute difference between Keras and TFLite is 0.027167, which is larger than the tolerance of 0.01.

This suggests that the TFLite float16 weight quantization path may not handle this attention softcap pattern with sufficient numerical consistency. Since float16 weight quantization is commonly used for on-device deployment, this may affect converted model correctness.

Expected behavior?

TFLite float16 weight quantization should produce results reasonably close to the original Keras fp32 model for this attention softcap pattern, or clearly document/warn if this pattern is expected to be numerically sensitive.

### Standalone code to reproduce the issue

```shell
#!/usr/bin/env python3
import os
import sys

os.environ.setdefault("CUDA_VISIBLE_DEVICES", "")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

try:
    import numpy as np
    import tensorflow as tf
except ImportError as e:
    print(f"missing dep: {e}", file=sys.stderr)
    sys.exit(2)

np.random.seed(42)

B, S, D, H = 2, 128, 64, 4
Dh, CAP, TOL = D // H, 5.0, 0.01

X = np.random.randn(B, S, D).astype(np.float32) * 2.0
Wq = np.random.randn(D, D).astype(np.float32)
Wk = np.random.randn(D, D).astype(np.float32)
Wv = np.random.randn(D, D).astype(np.float32)


class AttnSoftcap(tf.keras.Model):
    def __init__(self):
        super().__init__()
        self.wq = tf.Variable(Wq, trainable=False)
        self.wk = tf.Variable(Wk, trainable=False)
        self.wv = tf.Variable(Wv, trainable=False)

    @tf.function(input_signature=[tf.TensorSpec([B, S, D], tf.float32)])
    def call(self, x):
        q = tf.reshape(tf.matmul(x, self.wq), [B, S, H, Dh])
        k = tf.reshape(tf.matmul(x, self.wk), [B, S, H, Dh])
        v = tf.reshape(tf.matmul(x, self.wv), [B, S, H, Dh])

        q = tf.transpose(q, [0, 2, 1, 3])
        k = tf.transpose(k, [0, 2, 1, 3])
        v = tf.transpose(v, [0, 2, 1, 3])

        scores = tf.matmul(q, k, transpose_b=True) / (Dh ** 0.5)
        attn = tf.nn.softmax(tf.tanh(scores / CAP) * CAP, axis=-1)
        out = tf.matmul(attn, v)

        return tf.reshape(tf.transpose(out, [0, 2, 1, 3]), [B, S, D])


model = AttnSoftcap()
_ = model(tf.zeros([B, S, D], tf.float32))

keras_out = model(tf.constant(X)).numpy()
print(f"Keras: {keras_out.ravel()[:4]}")

converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_types = [tf.float16]
tflite_model = converter.convert()

interpreter = tf.lite.Interpreter(
    model_content=tflite_model,
    experimental_op_resolver_type=(
        tf.lite.experimental.OpResolverType.BUILTIN_WITHOUT_DEFAULT_DELEGATES
    ),
)

input_idx = interpreter.get_input_details()[0]["index"]
interpreter.resize_tensor_input(input_idx, list(X.shape))
interpreter.allocate_tensors()

output_idx = interpreter.get_output_details()[0]["index"]
interpreter.set_tensor(input_idx, X)
interpreter.invoke()

tflite_out = interpreter.get_tensor(output_idx)
print(f"TFLite:{tflite_out.ravel()[:4]}")

max_abs = float(np.max(np.abs(keras_out - tflite_out)))
print(f"max_abs={max_abs:.6f}")

if max_abs > TOL:
    print(f"BUG REPRODUCED: TFLite attention_logit_softcap (max_abs={max_abs:.6f})")
    sys.exit(0)

print("NOT reproduced")
sys.exit(1)
```

### Relevant log output

```shell
Keras: [2.414767 3.9525335 0.11160102 3.2639349 ]
TFLite:[2.4152098 3.9535334 0.11113904 3.2641575 ]
max_abs=0.027167
BUG REPRODUCED: TFLite attention_logit_softcap (max_abs=0.027167)
```

**Comment by aravindmurali22** [comment]
Hi @ALinrunrun ,

Thank you for taking the time to report this and for your patience.

TFLite has [transitioned](https://developers.googleblog.com/en/tensorflow-lite-is-now-litert/) to LiteRT, Google's primary runtime for on-device AI. All active development, feature updates, and performance enhancements have shifted  to LiteRT. We encourage you to try running your model with LiteRT. If you still encounter this issue, please open a new issue with the updated details in the [LiteRT Repository](https://github.com/google-ai-edge/LiteRT).

For more information and step-by-step guidance on updating your project, please refer to the [LiteRT Migration Guide](https://developers.google.com/edge/litert/migration).
Thank you again!
